### PR TITLE
Backup and restore API sessions to/from database

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Add save/restore of API sessions to the database to avoid a forced logout on FTL restarts. This updated the `pihole-FTL.db` database version to `15`. 

This PR add the new config option `webserver.api.session.restore` defaulting to `true` which some users may want to disable for security reasons if multiple users can read `pihole-FTL.db` and potentially "steal" sessions. Even when Pi-hole's API sessions are client IP isolated, such an attempt would still (theoretically) be possible for `localhost` users when local authorization is enforced otherwise. It should always be checked and questioned who should have write (and read!) access to the device you are running Pi-hole on.

This PR moves the existing config option `webserver.api.sessionTimeout` to `webserver.api.session.timeout`. The old config option will be automatically wiped from config files, the new one will be initialized with default values. As we are still in the beta testing phase, I skipped adding appropriate migration code to keep things simple.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Review by yubiuser — COMMENTED** [review]
You prepared a function `delete_all_sessions` but it seems you did not use it. I think after a successful restoration the function should be used to clean all sessions from the database.

__

Maybe this is a conceptual decision, but my understanding of the feature might be a bit different: you save all sessions to the database when they are created. However, I would only create those entries when we know we want to re-use them: right before we know `FTL` is going to be restarted (e.g. config change). In my view, there is no need to keep all those information stored all the time in the database, but only for as short as possible.

File: src/api/auth.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+*  API authentication prototypes
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+#ifndef AUTH_H
+#define AUTH_H
+
+// How many authenticated API clients are allowed simultaneously? [.]
+#define API_MAX_CLIENTS 16
```
[the remark this answers is not in the document]
Comment (DL6ER) [reply src/api/auth.h:15]: They are rejected

**Comment by DL6ER** [comment]
> You prepared a function `delete_all_sessions` but it seems you did not use it.

It is used when you change the password (or TOTP secret) - so something that invalidates all sessions at once.

---

> I would only create those entries when we know we want to re-use them

The idea was to keep the sessions also alive on unforeseen crashes (compare to how `iptables-restore` does it, too), but I do see you point in changing this into something that will only be there to backup a few milliseconds between (planned) restarts.

Should we also store them when FTL is `stop`ped and `start`ed later on manually (so not because of an API change)? FTL will anyway only restore still valid sessions but if users configures a timeout of, say, a month...

File: src/api/auth.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yubiuser) [reply src/api/auth.h:15]: Do we have any logging for that?

[the remark this answers is not in the document]
Comment (DL6ER) [reply src/api/auth.h:15]: I thought we had but looking at the code it just seems we reject the same way as if the password would be wrong. I will add this

**Comment by yubiuser** [comment]
> Should we also store them when FTL is stopped and started later on manually (so not because of an API change)?

In my view: no. The whole backup/restore is a nice thing to reduce annoyance when changing settings. I might even want to delete all sessions by manually trigger a restart.

**Comment by DL6ER** [comment]
Hmm, but I don't think a manual restart killing all sessions is an obvious workaround. My usecase is: I would like to preserve sessions across FTL recompilations and restarts. Maybe we need an `enum` instead of a simple `bool`

**Review by PromoFaux — APPROVED** [review]
Tested in a docker container.
Sessions successfully retained when shutting down the container gracefully (At which point FTL is gracefully shut down) :heavy_check_mark: 
Sessions not retained when force stopping the container (thus, FTL is killed without grace) :heavy_check_mark:

**Pull request merged**